[clang][dataflow] Merge `RecordValue`s with different locations correctly.

martinboehme opened the pull request:
Now that prvalue expressions map directly to values (see
https://reviews.llvm.org/D158977), it's no longer guaranteed that `RecordValue`s
associated with the same expression will always have the same storage location.

In other words, D158977 invalidated the assertion in `mergeDistinctValues()`.
The newly added test causes this assertion to fail without the other changes in
the patch.

This patch fixes the issue. However, the real fix will be to eliminate the
`StorageLocation` from `RecordValue` entirely.

kinu commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -121,18 +121,18 @@ static Value *mergeDistinctValues(QualType Type, Value &Val1,
+      MergedVal = &MergedEnv.create<RecordValue>(RecordVal1->getLoc());
+    else
+      // If the locations for the two records are different, need to create a
+      // completely new value.
```
Given that how cryptic when this could happen, it might be helpful to have a brief comment that both could happen depending on a subtle order of CFG and therefore how merge happens.

martinboehme replied:
Not sure exactly what you mean...

Whether the locations are the same or not doesn't really depend on the order in which we process CFG blocks. Instead, locations are guaranteed to be the same if the `RecordValue` is stored in `ExprToLoc`, and they are usually (but not always) different if the `RecordValue` is stored in `ExprToVal`.

The order in which we process CFG blocks affects whether we actually see different `RecordValue`s or not when the merge happens -- which makes merges difficult to trigger reliably from an integration test ([example](https://github.com/google/crubit/blob/3ab9d49fde5a59262903c589bb6963b8db4c0541/nullability/test/merge.cc#L342)). But I'm not sure how relevant that is to the situation here?

Anyway, my hope is that all of this will go away relatively soon because we're now working on eliminating the `RecordStorageLocation` from `RecordValue`.

kinu replied:
I think you're right, I wanted to see if some comment may help but it's a bit irrelevant here. Also given that it's going away I'm good with it, thanks!

kinu commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -121,18 +121,18 @@ static Value *mergeDistinctValues(QualType Type, Value &Val1,
+      // `RecordVal1` and `RecordVal2` may have different properties associated
+      // with them. Create a new `RecordValue` without any properties so that we
+      // soundly approximate both values. If a particular analysis needs to
+      // merge properties, it should do so in `DataflowAnalysis::merge()`.
```
'If a particular analysis needs to merge...' 
this part can be out of the if / else because the model's merge may do something for the else case too?

martinboehme replied:
I see what you're saying. However, if I put this part of the comment ("If a particular analysis needs to merge properties...") above the if statement, it would read a bit strangely -- because in the case where the locations aren't the same, it's clear we need to create a new `RecordValue` anyway, and this will of course remove any properties.

What I'm trying to explain here is that we need to create a new `RecordValue` even if the location is the same, because the properties may be different -- and that therefore the analysis needs to merge properties even in this case. Does that make sense?

Open to suggestions on how I could clarify this.

ymand replied:
Maybe expand a bit? e.g. "with the same location, but without any properties..."?

martinboehme replied:
I like the suggestion -- done!

kinu reviewed (commented):
Thanks, this LGTM (non-authoritative; but confirmed that it makes it go through with one of the real code I hit this)

martinboehme committed:
[clang][dataflow] Merge `RecordValue`s with different locations correctly.

Now that prvalue expressions map directly to values (see
https://reviews.llvm.org/D158977), it's no longer guaranteed that `RecordValue`s
associated with the same expression will always have the same storage location.

In other words, D158977 invalidated the assertion in `mergeDistinctValues()`.
The newly added test causes this assertion to fail without the other changes in
the patch.

This patch fixes the issue. However, the real fix will be to eliminate the
`StorageLocation` from `RecordValue` entirely.

martinboehme commented:
(Sorry, rebased to head inadvertently -- still getting used to the new process. Will use fixup commits in the future.)
